js/src/experimental/vercel/middleware.ts

```diff
@@ -0,0 +1,266 @@
+        async (_params: typeof params) => {
+          const result = await doGenerate();
+          const currentRunTree = getCurrentRunTree(true);
+          if (currentRunTree !== undefined) {
+            setUsageMetadataOnRunTree(result, currentRunTree);
+          }
+          return result;
+        },
+        {
+          ...lsConfig,
+          name: name ?? "ai.doGenerate",
+          run_type: "llm",
+          metadata: {
+            ls_model_name: modelId,
+            ai_sdk_method: "ai.doGenerate",
+            ...lsConfig?.metadata,
+          },
+          processInputs: (inputs) => {
+            const typedInputs = inputs as LanguageModelV2CallOptions;
+            return _formatTracedInputs(typedInputs);
+          },
+          processOutputs: (outputs) => {
+            const typedOutputs = outputs as Awaited<
+              ReturnType<typeof doGenerate>
+            >;
+            return _formatTracedOutputs(typedOutputs);
+          },
+        }
+      );
+      return traceableFunc(params);
+    },
+    wrapStream: async ({ doStream, params }) => {
+      const parentRunTree = getCurrentRunTree(true);
+      let runTree: RunTree | undefined;
+      if (
+        parentRunTree != null &&
+        typeof parentRunTree === "object" &&
+        typeof parentRunTree.createChild === "function"
+      ) {
+        runTree = parentRunTree?.createChild({
+          ...lsConfig,
+          name: name ?? "ai.doStream",
+          run_type: "llm",
+          metadata: {
+            ls_model_name: modelId,
+            ai_sdk_method: "ai.doStream",
+            ...lsConfig?.metadata,
+          },
+          inputs: _formatTracedInputs(params),
+        });
+      }
+
+      await runTree?.postRun();
+      try {
+        const { stream, ...rest } = await doStream();
+        const chunks: LanguageModelV2StreamPart[] = [];
+        const transformStream = new TransformStream({
+          async transform(chunk: LanguageModelV2StreamPart, controller) {
+            chunks.push(chunk);
+            controller.enqueue(chunk);
+          },
+
+          async flush() {
+            try {
+              const output = chunks.reduce(
+                (aggregated, chunk) => {
+                  if (chunk.type === "text-delta") {
+                    if (chunk.delta == null) {
+                      return aggregated;
+                    }
+                    return {
+                      ...aggregated,
+                      text: aggregated.text + chunk.delta,
+                    };
+                  } else if (chunk.type === "tool-call") {
+                    const matchingToolCall = aggregated.tool_calls.find(
+                      (call) => call.id === chunk.toolCallId
+                    );
+                    if (matchingToolCall != null) {
+                      return aggregated;
+                    }
+                    return {
+                      ...aggregated,
+                      tool_calls: [
```

Comment:
Is this correct? Looking through the input types of `_formatTracedOutputs` I would expect the return type to be `LanguageModelV2Message`, but this is not the case?

Reply:
Nvm chatted off bnad